### PR TITLE
Knative Service forwarder for EventsHub

### LINKED ISSUE #516: Allow Knative Service to be deployed along with EventsHub
When EventsHub is deployed as a Receiver for receiving events it would be good to have an option to deploy a Knative Service that would forward requests to the EventsHub Pod. Enabling the Knative Service forwarder should be done independently of Features, ideally through the Environment option.

This has a number of benefits:
* Allow for testing Knative Eventing and Serving together transparently
* Allow for reusing a huge number of already existing tests (their Features) from Eventing-related repositories (knative/eventing, knative-sandbox/eventing-kafka-broker, etc.) and test them together with Serving only by changing the environment
* Significantly reduce code duplication in downstream repositories - no need to recreate the whole Eventing test just to include Knative Serving

/assign

### PR DESCRIPTION
**Pull request by mgencur**
# Changes

- :gift:
- Add option for global environment to enable Knative Service Forwarder for EventsHub receiver:
```
ctx, env := global.Environment(eventshub.WithKnativeServiceForwarder)
```
- Implement forwarder that combines functionality from receiver and sender. It logs the events that are forwarded but doesn't record them (it doesn't send a k8s event for them).
- Tests which use EventsHub can stay untouched if they don't want to use the Knative Service forwarder. Tests which want to use it can re-use Features from other tests and only configure global environment like in the example below:
```
func TestPingSourceBrokerTriggerKsvc(t *testing.T) {
	t.Parallel()

	ctx, env := global.Environment(
		knative.WithKnativeNamespace(system.Namespace()),
		knative.WithLoggingConfig,
		knative.WithTracingConfig,
		k8s.WithEventListener,
                // vvv------ This line enables KSVC forwarder. Other lines are unchanged.
		eventshub.WithKnativeServiceForwarder,
		environment.Managed(t),
	)

        // features.SendsEventsWithSinkRef is reused (not changed)
	env.Test(ctx, t, features.SendsEventsWithSinkRef())
}
```

Other alternatives considered:
* Having EventsHub itself as Knative Service (vs. using the forwarder KSVC) - this approach has some drawbacks: The EventsHub records events and includes ObjectReference to the Pod that sends/receives the Event. This is problematic with Knative Serving because it might scale up and then there's no good way to get info about the concrete pod that is running.  It would have to be limited to `max-scale=1` which is kind of limiting for the desired usecases. This limitation is not there with the forwarder approach.
* Provide Knative Service as a separate StepFn - this limits the re-use of the whole Feature when someone want to test it together with Knative Service. It's no longer possible to re-use the whole Feature but it's required to re-implement it and add the KSVC in the scenario manually.

/kind <kind>

Fixes #516

**Release Note**

```release-note

```

**Docs**

```docs

```

### CODE REVIEW

**Review by pierDipi — COMMENTED** [review]
One problem is that we cannot depend on eventing (because eventing depends on rekt too), so we might need to move the forwarder in this repository and consume it from here in eventing

**Comment by mgencur** [comment]
> One problem is that we cannot depend on eventing (because eventing depends on rekt too), so we might need to move the forwarder in this repository and consume it from here in eventing

Uh. OK

**Comment by mgencur** [comment]
I may pull out the most important stuff from the eventing-forwarder and copy the code here. A lot of the stuff is not required here. Let's see what's the minimal impl that works here.

**Comment by mgencur** [comment]
> I may pull out the most important stuff from the eventing-forwarder and copy the code here. A lot of the stuff is not required here. Let's see what's the minimal impl that works here.

Actually, the Eventshub itself already has most of the functionality. I will try to re-use code from there and not depend on anything from Eventing.

**Comment by mgencur** [comment]
The current implementation of forwarder logs the forwarded events in this way:
```
{"level":"info","ts":"2023-05-02T08:55:00.368Z","logger":"eventshub.event logger","caller":"logger_vent/logger.go:26","msg":"Event: \n-- EventInfo --\n--- Kind: Received ---\n--- Event ---\nContext Attributes,\n  specversion: 1.0\n  type: dev.knative.sources.ping\n  source: /apis/v1/namespaces/test-sopikrdd/pingsources/pingsource-gbzgetvy\n  id: 893ff8c6-085f-46b5-9b3a-06028e2d24f8\n  time: 2023-05-02T08:55:00.287604272Z\nExtensions,\n  knativearrivaltime: 2023-05-02T08:55:00.29040074Z\n\n--- HTTP headers ---\n  X-B3-Spanid: 4fa16212ea1b1928\n  X-B3-Traceid: 0170963cc4b953629afd3438a5f3fdab\n  Content-Length: 0\n  Prefer: reply\n  Forwarded: for=10.128.2.13;proto=http\n  Accept-Encoding: gzip\n  Host: sink-kvnevsuz.test-sopikrdd.svc.cluster.local\n  K-Proxy-Request: activator\n  Traceparent: 00-0170963cc4b953629afd3438a5f3fdab-4fa16212ea1b1928-01\n  X-Forwarded-Proto: http\n  X-Forwarded-For: 10.128.2.13, 10.131.0.31\n  User-Agent: Go-http-client/1.1\n  X-Request-Id: e1ab22ed-2235-4e76-9423-ddcffaead355\n  X-B3-Sampled: 1\n\n--- Origin: '127.0.0.1:53498' ---\n--- Observer: 'sink-kvnevsuz' ---\n--- Time: 2023-05-02 08:55:00.367962994 +0000 UTC m=+23.086618826 ---\n--- Sequence: 0 ---\n--- Sent Id:  ' ---\n--------------------\n","commit":"c0e4ef7-dirty"}
{"level":"info","ts":"2023-05-02T08:55:00.398Z","logger":"eventshub.event logger","caller":"logger_vent/logger.go:26","msg":"Event: \n-- EventInfo --\n--- Kind: Sent ---\n--- Event ---\nContext Attributes,\n  specversion: 1.0\n  type: dev.knative.sources.ping\n  source: /apis/v1/namespaces/test-sopikrdd/pingsources/pingsource-gbzgetvy\n  id: 893ff8c6-085f-46b5-9b3a-06028e2d24f8\n  time: 2023-05-02T08:55:00.287604272Z\nExtensions,\n  knativearrivaltime: 2023-05-02T08:55:00.29040074Z\n\n--- HTTP headers ---\n  Ce-Knativearrivaltime: 2023-05-02T08:55:00.29040074Z\n  Ce-Id: 893ff8c6-085f-46b5-9b3a-06028e2d24f8\n  Ce-Specversion: 1.0\n  Ce-Type: dev.knative.sources.ping\n  Ce-Time: 2023-05-02T08:55:00.287604272Z\n  Ce-Source: /apis/v1/namespaces/test-sopikrdd/pingsources/pingsource-gbzgetvy\n\n--- Origin: 'sink-kvnevsuz' ---\n--- Observer: 'sink-kvnevsuz' ---\n--- Time: 2023-05-02 08:55:00.397931325 +0000 UTC m=+23.116587158 ---\n--- Sequence: 0 ---\n--- Sent Id:  '893ff8c6-085f-46b5-9b3a-06028e2d24f8 ---\n--------------------\n","commit":"c0e4ef7-dirty"}
{"level":"info","ts":"2023-05-02T08:55:00.398Z","logger":"eventshub.event logger","caller":"logger_vent/logger.go:26","msg":"Event: \n-- EventInfo --\n--- Kind: Response ---\n--- HTTP headers ---\n  Date: Tue, 02 May 2023 08:55:00 GMT\n  Content-Length: 0\n\n--- Body ---\n\n--- Status Code: 202 ---\n--- Origin: 'http://sink-kvnevsuz-rqrhixpu.test-sopikrdd.svc.cluster.local' ---\n--- Observer: 'sink-kvnevsuz' ---\n--- Time: 2023-05-02 08:55:00.398066529 +0000 UTC m=+23.116722360 ---\n--- Sequence: 0 ---\n--- Sent Id:  '893ff8c6-085f-46b5-9b3a-06028e2d24f8 ---\n--------------------\n","commit":"c0e4ef7-dirty"}
```
The forwarder also work with Tracing, this is a screenshot from Zipkin for an event that was sent in a test:
image

**Comment by mgencur** [comment]
Note: The Zipkin traces end at the Knative Service and don't continue further to eventshub Service. I will investigate this.

**Comment by mgencur** [comment]
Fixed the traces with the latest commit:
image

File: pkg/resources/knativeservice/ksvc.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package knativeservice
```
[the remark this answers is not in the document]
Comment (mgencur) [reply pkg/resources/knativeservice/ksvc.go:17]: Thanks. Pushed the changes.

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

**Pull request merged**